src/scanpy/preprocessing/_simple.py

```diff
@@ -720,14 +791,29 @@ def regress_out(
         regres = regressors_chunk[idx] if variable_is_categorical else regressors
         tasks.append(tuple((data_chunk, regres, variable_is_categorical)))
 
-    from joblib import Parallel, delayed
+    res = None
+    if not variable_is_categorical:
+        A = regres.to_numpy()
```

Comment:
It seems like this `regres` variable is [coming from the above loop](https://github.com/scverse/scanpy/pull/3284/files#diff-f9e0bdcdfc04622c421f0a5788bbba6ee0303750580d2915caba3239d799322fR788-R792) which could be bad.  I think the `A` variable should be clearly created from something that isn't declared in a loop, unless this is intentional (I assume not).  I would also move that loop into the `variable_is_categorical` block below.  That will make reasoning what `A` is easier anyway because it will need to be clearly declared as `regressor`

Reply:
More generally, I would split this and the above code into two blocks `variable_is_categorical` and `not variable_is_categorical`

Reply:
We also still have to check the logic and figure out a better task dispatch tree for the numpy shortcut. I also have to investigate How we format the array (C/F) for both versions

Reply:
I think this is out of scope for this update. To resolve the ideas you have would result in a complete rewrite of the regress_out function. There are more things that need to be updated for this to happen. So I would leave the logic as is for now.

Reply:
@Intron7 I think all that is required here is to use `regressors` instead of `regres`.  And then move the loop after this if-block into an `else` block.  It's unnecessary computation in the case that `not variable_is_categorical and np.linalg.det(A.T @ A) != 0`.

Previously, there were no `if-then` blocks around `variable_is_categorical` so seem pretty in-scope since we're adding those here.